### ISSUE TITLE
xpra ssh remote has become broken in some/many cases

### ISSUE THREAD

**Issue by karlkleinpaste** [Xpra-org/xpra#3626]
I hadn't updated in a couple weeks, until yesterday. Update has given me 4.4-10.r31917 on F35. Also trying to use Windows xpra, whatever the current normal release is, 4.3 something.

Attempting to start a remote xpra session, F35→F35, whether single terminal or desktop, gives this:
```
$ xpra start ssh://major/73 --attach=no --start-child=vterm
2022-09-23 17:56:38,829 Connected (version 2.0, client OpenSSH_8.7)
2022-09-23 17:56:38,959 Authentication (publickey) successful!
2022-09-23 17:56:39,627 SSH EOF on stderr of run-xpra
2022-09-23 17:56:39,633 Connection failed
```
No typical commentary about logs to be found under /run/user/1234/xpra/... and no session is started on the remote.
`vterm` is a simple script to pick what kind of terminal to start remotely. Mostly, it invokes `mate-terminal`.
Fortunately, I can just `ssh` to the remote to execute the start locally. After that, I can attach the session remotely as `ssh://host/D`.

Adding `-d ssh`:
```
$ xpra -d ssh start ssh://major/73 --attach=no --start-child=vterm
2022-09-23 17:43:52,003 debug enabled for xpra.net.ssh / ('network', 'ssh')
2022-09-23 17:43:52,058 debug enabled for xpra.scripts.parsing / ('ssh',)
2022-09-23 17:43:52,058 using paramiko ssh backend
2022-09-23 17:43:52,085 parsed user config '/home/karl/.ssh/config'
2022-09-23 17:43:52,085 116 hosts found
2022-09-23 17:43:52,091 got host config for 'major': {'port': '7887', 'user': 'karl', 'forwardagent': 'yes', 'forwardx11': 'yes', 'forwardx11trusted': 'yes', 'preferredauthentications': 'publickey,keyboard-interactive,password', 'gssapiauthentication': 'yes', 'pubkeyacceptedkeytypes': '+ssh-dss', 'hostname': 'major'}
2022-09-23 17:43:52,091 authentication modes=['none', 'agent', 'key', 'password']
2022-09-23 17:43:52,097 paramiko socket_connect: sockname=('10.0.0.202', 55602), peername=('10.0.0.203', 7887)
2022-09-23 17:43:52,097 SSH transport <paramiko.Transport at 0x3c074430 (unconnected)>
2022-09-23 17:43:52,097 starting thread (client mode): 0x3c074430
2022-09-23 17:43:52,097 Local version/idstring: SSH-2.0-paramiko_2.11.0
2022-09-23 17:43:52,110 Remote version/idstring: SSH-2.0-OpenSSH_8.7
2022-09-23 17:43:52,110 Connected (version 2.0, client OpenSSH_8.7)
2022-09-23 17:43:52,116 === Key exchange possibilities ===
2022-09-23 17:43:52,116 kex algos: curve25519-sha256, curve25519-sha256@libssh.org, ecdh-sha2-nistp256, ecdh-sha2-nistp384, ecdh-sha2-nistp521, diffie-hellman-group-exchange-sha256, diffie-hellman-group14-sha256, diffie-hellman-group16-sha512, diffie-hellman-group18-sha512
2022-09-23 17:43:52,116 server key: rsa-sha2-512, rsa-sha2-256, ecdsa-sha2-nistp256, ssh-ed25519
2022-09-23 17:43:52,116 client encrypt: aes256-gcm@openssh.com, chacha20-poly1305@openssh.com, aes256-ctr, aes128-gcm@openssh.com, aes128-ctr
2022-09-23 17:43:52,116 server encrypt: aes256-gcm@openssh.com, chacha20-poly1305@openssh.com, aes256-ctr, aes128-gcm@openssh.com, aes128-ctr
2022-09-23 17:43:52,116 client mac: hmac-sha2-256-etm@openssh.com, hmac-sha1-etm@openssh.com, umac-128-etm@openssh.com, hmac-sha2-512-etm@openssh.com, hmac-sha2-256, hmac-sha1, umac-128@openssh.com, hmac-sha2-512
2022-09-23 17:43:52,116 server mac: hmac-sha2-256-etm@openssh.com, hmac-sha1-etm@openssh.com, umac-128-etm@openssh.com, hmac-sha2-512-etm@openssh.com, hmac-sha2-256, hmac-sha1, umac-128@openssh.com, hmac-sha2-512
2022-09-23 17:43:52,116 client compress: none, zlib@openssh.com
2022-09-23 17:43:52,116 server compress: none, zlib@openssh.com
2022-09-23 17:43:52,116 client lang: <none>
2022-09-23 17:43:52,116 server lang: <none>
2022-09-23 17:43:52,116 kex follows: False
2022-09-23 17:43:52,116 === Key exchange agreements ===
2022-09-23 17:43:52,116 Kex: curve25519-sha256@libssh.org
2022-09-23 17:43:52,116 HostKey: ssh-ed25519
2022-09-23 17:43:52,116 Cipher: aes128-ctr
2022-09-23 17:43:52,116 MAC: hmac-sha2-256
2022-09-23 17:43:52,116 Compression: none
2022-09-23 17:43:52,116 === End of kex handshake ===
2022-09-23 17:43:52,164 kex engine KexCurve25519 specified hash_algo <built-in function openssl_sha256>
2022-09-23 17:43:52,164 Switch to new keys ...
2022-09-23 17:43:52,164 do_ssh_paramiko_connect_to(<paramiko.Transport at 0x3c074430 (cipher aes128-ctr, 128 bits) (connected; awaiting auth)>, 'major', 'karl', None, {'port': '7887', 'user': 'karl', 'forwardagent': 'yes', 'forwardx11': 'yes', 'forwardx11trusted': 'yes', 'preferredauthentications': 'publickey,keyboard-interactive,password', 'gssapiauthentication': 'yes', 'pubkeyacceptedkeytypes': '+ssh-dss', 'hostname': 'major'}, ['/home/karl/.ssh/id_ed25519', '/home/karl/.ssh/id_ecdsa', '/home/karl/.ssh/id_rsa', '/home/karl/.ssh/id_dsa'], {'display_name': 'ssh://major/73', 'cmdline': ['/usr/bin/xpra', '-d', 'ssh', 'start', 'ssh://major/73', '--attach=no', '--start-child=vterm'], 'type': 'ssh', 'proxy_command': ['_proxy_seamless'], 'exit_ssh': True, 'display': ':73', 'display_as_args': [':73', '--debug=ssh', '--start-child=vterm'], 'is_paramiko': True, 'agent': False, 'host': 'major', 'local': False, 'full_ssh': ['paramiko'], 'remote_xpra': ['xpra', '$XDG_RUNTIME_DIR/xpra/run-xpra', '/usr/local/bin/xpra', '~/.xpra/run-xpra', 'Xpra_cmd.exe']})
2022-09-23 17:43:52,164 Got EXT_INFO: {'server-sig-algs': b'ssh-ed25519,sk-ssh-ed25519@openssh.com,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,webauthn-sk-ecdsa-sha2-nistp256@openssh.com'}
2022-09-23 17:43:52,165 remote_server_key=MD5:29:85:47:61:d9:50:34:de:2e:4a:96:84:16:f2:86:9a
2022-09-23 17:43:52,208 HostKeys.load(/home/karl/.ssh/known_hosts) successful
2022-09-23 17:43:52,208 host keys=<paramiko.hostkeys.HostKeys object at 0x7f263c075ae0>
2022-09-23 17:43:52,209 ed25519 host key 'MD5:29:85:47:61:d9:50:34:de:2e:4a:96:84:16:f2:86:9a' OK for host 'major'
2022-09-23 17:43:52,209 starting authentication, authentication methods: ['none', 'agent', 'key', 'password']
2022-09-23 17:43:52,209 auth=none
2022-09-23 17:43:52,209 trying none authentication
2022-09-23 17:43:52,224 userauth is OK
2022-09-23 17:43:52,241 Authentication type (none) not permitted.
2022-09-23 17:43:52,241 Allowed methods: ['publickey', 'gssapi-keyex', 'gssapi-with-mic', 'password']
2022-09-23 17:43:52,242 auth_none()
Traceback (most recent call last):
  File "/usr/lib64/python3.10/site-packages/xpra/net/ssh.py", line 669, in auth_none
    transport.auth_none(username)
  File "/usr/lib/python3.10/site-packages/paramiko/transport.py", line 1501, in auth_none
    return self.auth_handler.wait_for_response(my_event)
  File "/usr/lib/python3.10/site-packages/paramiko/auth_handler.py", line 259, in wait_for_response
    raise e
paramiko.ssh_exception.BadAuthenticationType: Bad authentication type; allowed types: ['publickey', 'gssapi-keyex', 'gssapi-with-mic', 'password']
2022-09-23 17:43:52,242 auth=agent
2022-09-23 17:43:52,243 agent keys: (<paramiko.agent.AgentKey object at 0x7f263c11a350>, <paramiko.agent.AgentKey object at 0x7f263c11a500>, <paramiko.agent.AgentKey object at 0x7f263c11a4a0>, <paramiko.agent.AgentKey object at 0x7f263c11a560>, <paramiko.agent.AgentKey object at 0x7f263c11a5c0>, <paramiko.agent.AgentKey object at 0x7f263c11a620>)
2022-09-23 17:43:52,243 trying ssh-agent key 'MD5:87:b4:ad:51:e8:75:69:76:fe:4f:7e:25:4e:03:70:02'
2022-09-23 17:43:52,246 userauth is OK
2022-09-23 17:43:52,246 Finalizing pubkey algorithm for key of type 'ssh-rsa'
2022-09-23 17:43:52,246 Our pubkey algorithm list: ['rsa-sha2-512', 'rsa-sha2-256', 'ssh-rsa']
2022-09-23 17:43:52,246 Server-side algorithm list: ['ssh-ed25519', 'sk-ssh-ed25519@openssh.com', 'ssh-rsa', 'rsa-sha2-256', 'rsa-sha2-512', 'ssh-dss', 'ecdsa-sha2-nistp256', 'ecdsa-sha2-nistp384', 'ecdsa-sha2-nistp521', 'sk-ecdsa-sha2-nistp256@openssh.com', 'webauthn-sk-ecdsa-sha2-nistp256@openssh.com']
2022-09-23 17:43:52,246 Agreed upon 'rsa-sha2-512' pubkey algorithm
2022-09-23 17:43:52,258 Authentication (publickey) successful!
2022-09-23 17:43:52,259 authenticated using agent and key 'MD5:87:b4:ad:51:e8:75:69:76:fe:4f:7e:25:4e:03:70:02'
2022-09-23 17:43:52,259 will try to run xpra from: ['xpra', '$XDG_RUNTIME_DIR/xpra/run-xpra', '/usr/local/bin/xpra', '~/.xpra/run-xpra', 'Xpra_cmd.exe']
2022-09-23 17:43:52,259 paramiko_run_test_command(transport, command)
2022-09-23 17:43:52,259 [chan 0] Max packet in: 4096 bytes
2022-09-23 17:43:52,266 Received global request "hostkeys-00@openssh.com"
2022-09-23 17:43:52,266 Rejecting "hostkeys-00@openssh.com" global request from server.
2022-09-23 17:43:52,266 Debug msg: b'/home/karl/.ssh/authorized_keys:4: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding'
2022-09-23 17:43:52,311 [chan 0] Max packet out: 32768 bytes
2022-09-23 17:43:52,311 Secsh channel 0 opened.
2022-09-23 17:43:52,316 [chan find command] Sesch channel 0 request ok
2022-09-23 17:43:52,316 exec_command returned
2022-09-23 17:43:52,316 exit status is not ready yet, sleeping
2022-09-23 17:43:52,321 [chan find command] EOF received (find command)
2022-09-23 17:43:52,322 [chan find command] EOF sent (find command)
2022-09-23 17:43:52,326 exec_command('command')=0
2022-09-23 17:43:52,326 exec_command out=[]
2022-09-23 17:43:52,326 exec_command err=[]
2022-09-23 17:43:52,326 paramiko_run_test_command(transport, command -v xpra)
2022-09-23 17:43:52,326 [chan 1] Max packet in: 4096 bytes
2022-09-23 17:43:52,369 [chan 1] Max packet out: 32768 bytes
2022-09-23 17:43:52,369 Secsh channel 1 opened.
2022-09-23 17:43:52,375 [chan find command -v xpra] Sesch channel 1 request ok
2022-09-23 17:43:52,375 exec_command returned
2022-09-23 17:43:52,375 exit status is not ready yet, sleeping
2022-09-23 17:43:52,385 exit status is not ready yet, sleeping
2022-09-23 17:43:52,386 [chan find command -v xpra] EOF received (find command -v xpra)
2022-09-23 17:43:52,386 [chan find command -v xpra] EOF sent (find command -v xpra)
2022-09-23 17:43:52,396 exec_command('command -v xpra')=0
2022-09-23 17:43:52,396 exec_command out=['/usr/bin/xpra\n']
2022-09-23 17:43:52,396 exec_command err=[]
2022-09-23 17:43:52,396 adding xpra_cmd='/usr/bin/xpra'
2022-09-23 17:43:52,396 cmd(['_proxy_seamless'], [':73', '--debug=ssh', '--start-child=vterm'])="/usr/bin/xpra" "_proxy_seamless" ":73" "--debug=ssh" "--start-child=vterm"
2022-09-23 17:43:52,396 trying to open SSH session, window-size=134217727, timeout=60
2022-09-23 17:43:52,396 [chan 2] Max packet in: 4096 bytes
2022-09-23 17:43:52,438 [chan 2] Max packet out: 32768 bytes
2022-09-23 17:43:52,438 Secsh channel 2 opened.
2022-09-23 17:43:52,438 paramiko agent_option=False
2022-09-23 17:43:52,438 channel exec_command("/usr/bin/xpra" "_proxy_seamless" ":73" "--debug=ssh" "--start-child=vterm"
2022-09-23 17:43:52,443 [chan run-xpra] Sesch channel 2 request ok
2022-09-23 17:43:52,535 [chan run-xpra] EOF received (run-xpra)
2022-09-23 17:43:52,535 SSH EOF on stderr of run-xpra
2022-09-23 17:43:52,540 [chan run-xpra] EOF sent (run-xpra)
2022-09-23 17:43:52,541 Connection failed
```
There are similar problems when trying merely to attach the Windows shadow server. After starting the shadow server on a nearby Win10 box, attempting to connect from F35 gives this failure:
```
$ xpra attach ssh://francine@espresso:14500
2022-09-23 18:07:16,578 Xpra GTK3 X11 client version 4.4-r31917 (g27951736b) beta 64-bit
2022-09-23 18:07:16,599  running on Linux Fedora 35 ThirtyFive
2022-09-23 18:07:16,600  window manager is 'compiz'
2022-09-23 18:07:17,291 Error: failed to query sound subsystem:
2022-09-23 18:07:17,291  query did not return any data
2022-09-23 18:07:17,401 created unix domain socket '/run/user/5271/xpra/clients/godiva.kleinpaste.org-341506'
2022-09-23 18:07:17,419 Error importing nvjpeg decoder (dec_nvjpeg)
2022-09-23 18:07:17,419  libnvjpeg.so.11: cannot open shared object file: No such file or directory
2022-09-23 18:07:17,446 OpenGL_accelerate module loaded
2022-09-23 18:07:17,451 Using accelerated ArrayDatatype
2022-09-23 18:07:17,620 OpenGL enabled with NVIDIA GeForce RTX 2070 with Max-Q Design/PCIe/SSE2
2022-09-23 18:07:17,750 Connected (version 2.0, client paramiko_2.11.0)
2022-09-23 18:07:17,827 Authentication (none) successful!
Warning: failed to connect:
 connection failed: Channel closed.
2022-09-23 18:07:17,849 removing unix domain socket '/run/user/5271/xpra/clients/godiva.kleinpaste.org-341506'
```
Just a blind closure, failed, no explanation.

**Comment by totaam** [comment]
Sorry about that, this is very likely to have been caused by 28700308f2de43b147bd9653765f0c2b17e7afc1 using `_proxy_seamless` instead of just `_proxy` - which works with newer versions but not with 4.3.x. Will fix.

**Comment by karlkleinpaste** [comment]
Does this explain the F35→F35 (4.4-10.r31917 at both ends) startup problem as well?

**Comment by totaam** [comment]
> Does this explain the F35→F35 (4.4-10.r31917 at both ends) startup problem as well?

No, it does not. There must be something else to it. Will fix.

**Comment by totaam** [comment]
Sorry about the breakage and thanks for the ticket!

> `xpra start ssh://major/73 --attach=no --start-child=vterm`

This now correctly calls `_proxy_start` which is backwards compatible with older xpra versions:
```shell
cmd(['_proxy_start'], [':10', '--debug=ssh', '--start-child=xterm'])=\
    "/usr/bin/xpra" "_proxy_start" ":10" "--debug=ssh" "--start-child=xterm"
```
And thanks to another fix (e83dd8dde69dd4a7e890b157589604edcaedce4c), this now also correctly starts the new `seamless` server instance.

**Comment by karlkleinpaste** [comment]
Updated to 4.4-10.r31959.
I appreciate the effort, and the F35→F35 problem is solved. However, I still cannot reach the Windows shadow server, same symptom as before.
```
$ xpra attach ssh://francine@espresso:14500
2022-09-25 04:00:42,368 Xpra GTK3 X11 client version 4.4-r31959 (g7209254d2) beta 64-bit
2022-09-25 04:00:42,390  running on Linux Fedora 35 ThirtyFive
2022-09-25 04:00:42,390  window manager is 'compiz'
2022-09-25 04:00:43,077 GStreamer loaded
2022-09-25 04:00:43,160 created unix domain socket '/run/user/5271/xpra/clients/godiva.kleinpaste.org-797179'
2022-09-25 04:00:43,209 OpenGL_accelerate module loaded
2022-09-25 04:00:43,212 Using accelerated ArrayDatatype
2022-09-25 04:00:43,362 OpenGL enabled with NVIDIA GeForce RTX 2070 with Max-Q Design/PCIe/SSE2
2022-09-25 04:00:43,426 Connected (version 2.0, client paramiko_2.11.0)
2022-09-25 04:00:43,564 Authentication (none) successful!
Warning: failed to connect:
 connection failed: Channel closed.
2022-09-25 04:00:43,584 removing unix domain socket '/run/user/5271/xpra/clients/godiva.kleinpaste.org-797179'
```
Adding `-d ssh` to this session, in hopes it will help:
```
$ xpra -d ssh attach ssh://francine@espresso:14500
2022-09-25 04:02:32,701 Xpra GTK3 X11 client version 4.4-r31959 (g7209254d2) beta 64-bit
2022-09-25 04:02:32,723  running on Linux Fedora 35 ThirtyFive
2022-09-25 04:02:32,724  window manager is 'compiz'
2022-09-25 04:02:33,389 GStreamer loaded
2022-09-25 04:02:33,438 debug enabled for xpra.net.ssh / ('network', 'ssh')
2022-09-25 04:02:33,502 created unix domain socket '/run/user/5271/xpra/clients/godiva.kleinpaste.org-797680'
2022-09-25 04:02:33,549 OpenGL_accelerate module loaded
2022-09-25 04:02:33,552 Using accelerated ArrayDatatype
2022-09-25 04:02:33,736 OpenGL enabled with NVIDIA GeForce RTX 2070 with Max-Q Design/PCIe/SSE2
2022-09-25 04:02:33,738 debug enabled for xpra.scripts.parsing / ('ssh',)
2022-09-25 04:02:33,738 using paramiko ssh backend
2022-09-25 04:02:33,742 parsed user config '/home/karl/.ssh/config'
2022-09-25 04:02:33,742 116 hosts found
2022-09-25 04:02:33,748 got host config for 'espresso': {'user': 'francine', 'forwardx11': 'no', 'forwardx11trusted': 'no', 'port': '7887', 'forwardagent': 'yes', 'preferredauthentications': 'publickey,keyboard-interactive,password', 'gssapiauthentication': 'yes', 'pubkeyacceptedkeytypes': '+ssh-dss', 'hostname': 'espresso'}
2022-09-25 04:02:33,748 authentication modes=['none', 'agent', 'key', 'password']
2022-09-25 04:02:33,755 paramiko socket_connect: sockname=('10.0.0.202', 33622), peername=('10.0.0.29', 14500)
2022-09-25 04:02:33,755 SSH transport <paramiko.Transport at 0xd4d6ea10 (unconnected)>
2022-09-25 04:02:33,756 starting thread (client mode): 0xd4d6ea10
2022-09-25 04:02:33,756 Local version/idstring: SSH-2.0-paramiko_2.11.0
2022-09-25 04:02:33,803 Remote version/idstring: SSH-2.0-paramiko_2.11.0
2022-09-25 04:02:33,803 Connected (version 2.0, client paramiko_2.11.0)
2022-09-25 04:02:33,808 === Key exchange possibilities ===
2022-09-25 04:02:33,808 kex algos: curve25519-sha256@libssh.org, ecdh-sha2-nistp256, ecdh-sha2-nistp384, ecdh-sha2-nistp521, diffie-hellman-group16-sha512, diffie-hellman-group-exchange-sha256, diffie-hellman-group14-sha256, diffie-hellman-group-exchange-sha1, diffie-hellman-group14-sha1, diffie-hellman-group1-sha1
2022-09-25 04:02:33,808 server key: rsa-sha2-512, rsa-sha2-256, ssh-rsa
2022-09-25 04:02:33,808 client encrypt: aes128-ctr, aes192-ctr, aes256-ctr, aes128-cbc, aes192-cbc, aes256-cbc, 3des-cbc
2022-09-25 04:02:33,808 server encrypt: aes128-ctr, aes192-ctr, aes256-ctr, aes128-cbc, aes192-cbc, aes256-cbc, 3des-cbc
2022-09-25 04:02:33,808 client mac: hmac-sha2-256, hmac-sha2-512, hmac-sha2-256-etm@openssh.com, hmac-sha2-512-etm@openssh.com, hmac-sha1, hmac-md5, hmac-sha1-96, hmac-md5-96
2022-09-25 04:02:33,808 server mac: hmac-sha2-256, hmac-sha2-512, hmac-sha2-256-etm@openssh.com, hmac-sha2-512-etm@openssh.com, hmac-sha1, hmac-md5, hmac-sha1-96, hmac-md5-96
2022-09-25 04:02:33,808 client compress: none
2022-09-25 04:02:33,808 server compress: none
2022-09-25 04:02:33,808 client lang: <none>
2022-09-25 04:02:33,808 server lang: <none>
2022-09-25 04:02:33,808 kex follows: False
2022-09-25 04:02:33,808 === Key exchange agreements ===
2022-09-25 04:02:33,808 Kex: curve25519-sha256@libssh.org
2022-09-25 04:02:33,808 HostKey: rsa-sha2-512
2022-09-25 04:02:33,808 Cipher: aes128-ctr
2022-09-25 04:02:33,808 MAC: hmac-sha2-256
2022-09-25 04:02:33,808 Compression: none
2022-09-25 04:02:33,808 === End of kex handshake ===
2022-09-25 04:02:33,822 kex engine KexCurve25519 specified hash_algo <built-in function openssl_sha256>
2022-09-25 04:02:33,827 Switch to new keys ...
2022-09-25 04:02:33,827 do_ssh_paramiko_connect_to(<paramiko.Transport at 0xd4d6ea10 (cipher aes128-ctr, 128 bits) (connected; awaiting auth)>, 'espresso', 'francine', None, {'user': 'francine', 'forwardx11': 'no', 'forwardx11trusted': 'no', 'port': '7887', 'forwardagent': 'yes', 'preferredauthentications': 'publickey,keyboard-interactive,password', 'gssapiauthentication': 'yes', 'pubkeyacceptedkeytypes': '+ssh-dss', 'hostname': 'espresso'}, ['/home/karl/.ssh/id_ed25519', '/home/karl/.ssh/id_ecdsa', '/home/karl/.ssh/id_rsa', '/home/karl/.ssh/id_dsa'], {'display_name': 'ssh://francine@espresso:14500', 'cmdline': ['/usr/bin/xpra', '-d', 'ssh', 'attach', 'ssh://francine@espresso:14500'], 'type': 'ssh', 'proxy_command': ['_proxy'], 'exit_ssh': True, 'display': None, 'display_as_args': [], 'is_paramiko': True, 'agent': False, 'username': 'francine', 'host': 'espresso', 'local': False, 'port': 14500, 'full_ssh': ['paramiko'], 'remote_xpra': ['xpra', '$XDG_RUNTIME_DIR/xpra/run-xpra', '/usr/local/bin/xpra', '~/.xpra/run-xpra', 'Xpra_cmd.exe']})
2022-09-25 04:02:33,827 Got EXT_INFO: {'server-sig-algs': b'ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-dss'}
2022-09-25 04:02:33,827 remote_server_key=MD5:4f:94:0c:70:0d:ee:69:6a:75:88:5a:3d:08:14:77:b0
2022-09-25 04:02:33,871 HostKeys.load(/home/karl/.ssh/known_hosts) successful
2022-09-25 04:02:33,871 host keys=<paramiko.hostkeys.HostKeys object at 0x7efbd4d6fd60>
2022-09-25 04:02:33,873 rsa host key 'MD5:4f:94:0c:70:0d:ee:69:6a:75:88:5a:3d:08:14:77:b0' OK for host 'espresso'
2022-09-25 04:02:33,873 starting authentication, authentication methods: ['none', 'agent', 'key', 'password']
2022-09-25 04:02:33,873 auth=none
2022-09-25 04:02:33,873 trying none authentication
2022-09-25 04:02:33,878 userauth is OK
2022-09-25 04:02:33,883 Authentication (none) successful!
2022-09-25 04:02:33,884 will try to run xpra from: ['xpra', '$XDG_RUNTIME_DIR/xpra/run-xpra', '/usr/local/bin/xpra', '~/.xpra/run-xpra', 'Xpra_cmd.exe']
2022-09-25 04:02:33,884 paramiko_run_test_command(transport, command)
2022-09-25 04:02:33,884 [chan 0] Max packet in: 4096 bytes
2022-09-25 04:02:33,890 [chan 0] Max packet out: 32768 bytes
2022-09-25 04:02:33,890 Secsh channel 0 opened.
2022-09-25 04:02:33,895 [chan find command] EOF sent (find command)
Warning: failed to connect:
 connection failed: Channel closed.
2022-09-25 04:02:33,908 removing unix domain socket '/run/user/5271/xpra/clients/godiva.kleinpaste.org-797680'
2022-09-25 04:02:33,950 [chan find command] EOF received (find command)

```

**Comment by totaam** [comment]
I think the MS Windows problem is caused by this:
```
paramiko_run_test_command(transport, command)
```
That's yet more breakage caused by #3215.

@karlkleinpaste which ssh server are you using and how do I set it up for testing?

**Comment by karlkleinpaste** [comment]
Cygwin/X.
https://gist.githubusercontent.com/roxlu/5038729/raw/4e8d02d1d8aa0acf3f17a55855563038ac50309b/install_cygwin_sshd.txt
The core content from those instructions can be summarized more succinctly (because they appear to be rather out of date):
* Get setup from https://x.cygwin.com/
* Install basics.
* Install openssh + cygrunsrv.
* Open cygwin terminal as adminstrator.
* ssh-host-config, accept all defaults.
* Ignore everything to do with special accounts, not needed.
* I don't remember if I had to open the port in the firewall, but the instructions for doing so are solid, if needed.

**Comment by karlkleinpaste** [comment]
I'm just curious about something.
The spec I'm using to reach the WIndows shadow server is ssh-based, but includes :14500. I notice that when xpra is trying to connect, it's doing so to this port, not the ssh port. And it then proceeds to announce successful "none" auth:
```
2022-09-25 04:32:45,998 starting authentication, authentication methods: ['none', 'agent', 'key', 'password']
2022-09-25 04:32:45,998 auth=none
2022-09-25 04:32:45,999 trying none authentication
2022-09-25 04:32:46,003 userauth is OK
2022-09-25 04:32:46,010 Authentication (none) successful!
```
Do you even need an actual ssh server for this?

**Comment by totaam** [comment]
> it's doing so to this port, not the ssh port

Ah, so it's using the builtin ssh server.
What is your shadow command line?

**Comment by karlkleinpaste** [comment]
I'm not aware that I've ever set up the builtin server, so unless it's configured by default, it's not running.
I am definitely running cygwin/x sshd and can login remotely using it.

Shadow server was started using the native tool: started xpra, got window with 4 big buttons, bottom left is "Shadow." Hit that. By looking around xpra source, that evidently invokes `xpra shadow --bind-tcp=0.0.0.0:14500,auth=sys,ssl-cert=auto` (from `packaging/MSWindows/xpra.iss`).

**Comment by totaam** [comment]
> I'm not aware that I've ever set up the builtin server, so unless it's configured by default, it's not running.

It is.
If you have `tcp` sockets then they can be upgraded to `ssh`: https://github.com/Xpra-org/xpra/blob/master/docs/Network/SSH.md#builtin-ssh-server

---

And down the rabbit hole I ended up..
First up, I was wrong about this:
> I think there's a bug there - `none` auth should fail

It's fine. By default, we don't do `ssh` authentication, we use xpra's own protocol for that *after* the ssh handshake.

It wasn't all caused by #3215, but that change certainly made things more _interesting_.
* the builtin ssh server has to simulate running `command` so that we can see if it is available 2a82813cb3509f16d0f942d518b40ecdfbdb949a
* problem is that `command` means something completely different on Microsoft Windows (`command.exe`) and this causes the ssh connections to terminate on that platform.. so we have to run the OS detection **before** we can run `command` 98cc320cf730993bb7e4c3d837edf2e55c3ce2c2 : the OS detection was made more reliable using `echo %OS%` (which should run on any OS) instead of `ver` (a command by that name could exist on a Posix OS)
* found a recent bug caused by 55913f39bdbeee1af23a25a6c61d792dd107f8d2, fixed in bd707c5eb7f52028d9eef79a0dce70b442b8017c which prevented ssh attach from working
* support `echo` in the builtin ssh server and add ssh banner: eba6c2a1b90ee97d0d0af01d8130ef0f96379f68
* and the really tough one: 1b017a365dcbbca23cd6815d128f2cbbb3f5421a undocumented paramiko API usage to support multiple exec requests (this is not a new bug, but the switch to using `command` made this a showstopper)
* fake batch script execution: 2fc9b2211b5df7697f56d1d27fb7fde085b217f0

Minor bits:
* the new OS detection could fail when running against a buggy server: 2f84cab36c35b6102c3ce06ed902214042245c5f
* log file splatter on ssh socket close: 9f9ba1c2c51e0386823d0c6888f297cc538b8ed3
* cosmetic tweaks: c56a9c82cc38fa8e23507d5c589d27006143f97a, aaece31145ca6402c32e02d24e627f9749e6082c, abdc5aa2829138ecf75d7ee93047da105e6e87e7, a44ee9ada5dd3fe64bc950b4836467fa9aa82a55
* refactoring: 0f1d62444dfcf7123f7b8b5dac38254c2647d50a
* prefer the first matching xpra command path specified: 569faed4acac22a26a3b1f256a7a4d7e7d9e44c5
* simplify c1ab2e4a00f12af3f420c2f8abc10b641a840a23
* regression introduced when failing to load a host key file: ad6f3432603e53b2c8a24325bf89b0cfce3a4f4d
* fix for loading `dsa` host keys: d7ac55a4193137a25b732466ee59e76f35da82e6

---

**Finally**. Tested with the builtin openssh server on Fedora, the openssh server on MS Windows installed via the appstore (`net start sshd`), `sshd` from MSYS2 (via `cygrunsrv`) and the builtin ssh server on all other platforms.

MS Windows builds uploaded, the other builds will follow.

**Comment by karlkleinpaste** [comment]
This has fixed up the details mentioned specifically within, so I'm closing it.
I discovered a new bug as a result, but it's different enough that I've filed a separate bug.
